Issue copying state of type State<AutoDiffX>

I am attempting to do the following:

If I add the following function to ContinuousState:

```
  void SetFrom(const ContinuousState<AutoDiffXd>& other) {
    DRAKE_DEMAND(size() == other.size());
    DRAKE_DEMAND(get_generalized_position().size() ==
        other.get_generalized_position().size());
    DRAKE_DEMAND(get_generalized_velocity().size() ==
        other.get_generalized_velocity().size());
    DRAKE_DEMAND(get_misc_continuous_state().size() ==
        other.get_misc_continuous_state().size());
    state_->SetFromVector(other.CopyToVector());
  }

```
and the following function to State:
```
  void SetFrom(const State<AutoDiffXd>& other) {
    continuous_state_->SetFrom(*other.get_continuous_state());
    discrete_state_->SetFrom(*other.get_discrete_state());
    abstract_state_->CopyFrom(*other.get_abstract_state());
  }

```

I get the following syntax error when attempting to compile a Simulator<AutoDiffXd>:

```
In file included from /home/drum/drake/drake/systems/framework/continuous_state.cc:1:0:
/home/drum/drake/drake/../drake/systems/framework/continuous_state.h: In instantiation of ‘void drake::systems::ContinuousState<T>::SetFrom(const drake::systems::ContinuousState<Eigen::AutoDiffScalar<Eigen::Matrix<double, -1, 1> > >&) [with T = double]’:
/home/drum/drake/drake/systems/framework/continuous_state.cc:6:16:   required from here
/home/drum/drake/drake/../drake/systems/framework/continuous_state.h:168:5: error: no matching function for call to ‘drake::systems::VectorBase<double>::SetFromVector(drake::VectorX<Eigen::AutoDiffScalar<Eigen::Matrix<double, -1, 1> > >)’
     state_->SetFromVector(other.CopyToVector()); 
```

This issue will cause the Simulator<AutoDiffXd> instantiation to be disabled on the unrestricted updates PR (#4382).

--- a/drake/systems/framework/continuous_state.cc
+++ b/drake/systems/framework/continuous_state.cc
@@ -1,9 +1,13 @@
 #include "drake/systems/framework/continuous_state.h"
 
+#include "drake/common/autodiff_overloads.h"
+#include "drake/common/eigen_autodiff_types.h"
+
 namespace drake {
 namespace systems {
 
 template class ContinuousState<double>;
+template class ContinuousState<AutoDiffXd>;
 
 }  // namespace systems
 }  // namespace drake

--- a/drake/systems/framework/continuous_state.h
+++ b/drake/systems/framework/continuous_state.h
@@ -145,15 +145,17 @@ class ContinuousState {
   }
 
 
+  /// Copies the values from another ContinuousState of the same scalar type
+  /// into this State.
+  void CopyFrom(const ContinuousState<T>& other) {
+    SetFromGeneric(other);
+  }
+
+  /// Initializes this ContinuousState (regardless of scalar type) from a
+  /// State<double>. All scalar types in Drake must support initialization from
+  /// doubles.
   void SetFrom(const ContinuousState<double>& other) {
-    DRAKE_DEMAND(size() == other.size());
-    DRAKE_DEMAND(get_generalized_position().size() ==
-                 other.get_generalized_position().size());
-    DRAKE_DEMAND(get_generalized_velocity().size() ==
-                 other.get_generalized_velocity().size());
-    DRAKE_DEMAND(get_misc_continuous_state().size() ==
-                 other.get_misc_continuous_state().size());
-    SetFromVector(other.CopyToVector().template cast<T>());
+    SetFromGeneric(other);
   }
 
   /// Sets the entire continuous state vector from an Eigen expression.
@@ -172,6 +174,18 @@ class ContinuousState {
   ContinuousState& operator=(ContinuousState&& other) = delete;
 
  private:
+  template <typename U>
+  void SetFromGeneric(const ContinuousState<U>& other) {
+    DRAKE_DEMAND(size() == other.size());
+    DRAKE_DEMAND(get_generalized_position().size() ==
+        other.get_generalized_position().size());
+    DRAKE_DEMAND(get_generalized_velocity().size() ==
+        other.get_generalized_velocity().size());
+    DRAKE_DEMAND(get_misc_continuous_state().size() ==
+        other.get_misc_continuous_state().size());
+    SetFromVector(other.CopyToVector().template cast<T>());
+  }
+
   // The entire continuous state vector.  May or may not own the underlying
   // data.
   std::unique_ptr<VectorBase<T>> state_;

--- a/drake/systems/framework/state.cc
+++ b/drake/systems/framework/state.cc
@@ -1,4 +1,13 @@
-// For now, this is an empty .cc file that only serves to confirm
-// state.h is a stand-alone header.
-
 #include "drake/systems/framework/state.h"
+
+#include "drake/common/autodiff_overloads.h"
+#include "drake/common/eigen_autodiff_types.h"
+
+namespace drake {
+namespace systems {
+
+template class State<double>;
+template class State<AutoDiffXd>;
+
+}  // namespace systems
+}  // namespace drake

--- a/drake/systems/framework/state.h
+++ b/drake/systems/framework/state.h
@@ -63,6 +63,16 @@ class State {
     return abstract_state_.get();
   }
 
+  /// Copies the values from another State of the same scalar type into this
+  /// State.
+  void CopyFrom(const State<T>& other) {
+    continuous_state_->CopyFrom(*other.get_continuous_state());
+    discrete_state_->CopyFrom(*other.get_discrete_state());
+    abstract_state_->CopyFrom(*other.get_abstract_state());
+  }
+
+  /// Initializes this state (regardless of scalar type) from a State<double>.
+  /// All scalar types in Drake must support initialization from doubles.
   void SetFrom(const State<double>& other) {
     continuous_state_->SetFrom(*other.get_continuous_state());
     discrete_state_->SetFrom(*other.get_discrete_state());

--- a/drake/systems/framework/test/continuous_state_test.cc
+++ b/drake/systems/framework/test/continuous_state_test.cc
@@ -91,6 +91,19 @@ TEST_F(ContinuousStateTest, MoreVelocityThanPositionVariables) {
       std::out_of_range);
 }
 
+TEST_F(ContinuousStateTest, CopyFrom) {
+  // Create a zero-initialized continuous state, with the same dimensions as
+  // the continuous state in the fixture.
+  ContinuousState<int> next_state(BasicVector<int>::Make({0, 0, 0, 0}),
+                                  kPositionLength, kVelocityLength,
+                                  kMiscLength);
+  next_state.CopyFrom(*continuous_state_);
+  EXPECT_EQ(1, next_state[0]);
+  EXPECT_EQ(2, next_state[1]);
+  EXPECT_EQ(3, next_state[2]);
+  EXPECT_EQ(4, next_state[3]);
+}
+
 }  // namespace
 }  // namespace systems
 }  // namespace drake

--- a/drake/systems/framework/test/leaf_context_test.cc
+++ b/drake/systems/framework/test/leaf_context_test.cc
@@ -282,6 +282,20 @@ TEST_F(LeafContextTest, Clone) {
   EXPECT_EQ(1.0, context_.get_numeric_parameter(0)->GetAtIndex(0));
 }
 
+// Tests that the State can be copied from another State.
+TEST_F(LeafContextTest, CopyStateFrom) {
+  std::unique_ptr<Context<double>> clone = context_.Clone();
+  (*clone->get_mutable_continuous_state())[0] = 81.0;
+  (*clone->get_mutable_discrete_state(0))[0] = 243.0;
+  clone->get_mutable_abstract_state<int>(0) = 729;
+
+  context_.get_mutable_state()->CopyFrom(clone->get_state());
+
+  EXPECT_EQ(81.0, (*context_.get_continuous_state())[0]);
+  EXPECT_EQ(243.0, (*context_.get_discrete_state(0))[0]);
+  EXPECT_EQ(729, context_.get_abstract_state<int>(0));
+}
+
 // Tests that a LeafContext<AutoDiffXd> can be initialized from a
 // Leafcontext<double>.
 TEST_F(LeafContextTest, SetTimeStateAndParametersFrom) {

